### DISCUSSION TITLE
Retrieve list of columns for table with traverser API

### DISCUSSION THREAD

**Discussion by henry99923123** [jOOQ/jOOQ#15709] [Q&A]
Consider this query:

`context.select().from(Author.AUTHOR);`

With the new Traverser API. How could one retrieve all columns that will be resolved in the generated SELECT *? I couldn't find a method in the context.meta()-API for this.

**Comment by lukaseder** [comment]
The `Traverser` API won't give you the resolved columns, because that's not what it's supposed to do. It's supposed to traverse the query expression tree as it was created, not as it will be rendered. For example, when your query uses syntax that requires emulation (e.g. the `QUALIFY` clause), the `Traverser` API won't see that emulation. The main reasons are:

- That's not what people want when traversing
- A traverser is independent of the `SQLDialect` and all other rendering `Settings`. You can traverse any `QueryPart` type, not just `Attachable` ones.

To access the resolved projection, you can either use a `VisitListener` (an older SPI that allows for listening to all `Context::visit` events of all actually rendered `QueryPart` objects, including emulations), or just use a very historic API: `Select::getSelect`. An API that should have remained internal, but has been public for ages, so it probably won't ever be made internal.
